File: C++/2-Medium/00156._上下翻转二叉树/00156._上下翻转二叉树.cpp
#include <iostream>
using namespace std;

// 定义二叉树节点结构
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

// 上下翻转二叉树的函数
TreeNode* upsideDownBinaryTree(TreeNode* root) {
    // 边界条件：如果节点为空，或者没有左子节点，直接返回当前节点
    if (!root || !root->left) {
        return root;
    }
    // 递归翻转左子树，并获取新的根节点
    TreeNode* newRoot = upsideDownBinaryTree(root->left);
    // 将当前节点的左子节点指向当前节点的右兄弟节点，实现翻转
    root->left->left = root->right;
    // 将当前节点的左子节点的右子节点指向当前节点，完成翻转
    root->left->right = root;
    // 避免循环引用，将当前节点的左右子节点设为null
    root->left = NULL;
    root->right = NULL;
    // 返回新的根节点
    return newRoot;
}

// 辅助函数，用于创建示例二叉树
TreeNode* createExampleTree() {
    TreeNode* root = new TreeNode(1);
    root->left = new TreeNode(2);
    root->right = new TreeNode(3);
    root->left->left = new TreeNode(4);
    root->left->right = new TreeNode(5);
    return root;
}

// 主函数
int main() {
    // 创建示例二叉树
    TreeNode* root = createExampleTree();
    // 进行上下翻转
    TreeNode* newRoot = upsideDownBinaryTree(root);
    // 输出结果，此处省略输出代码，实际中可以遍历打印树来验证结果
    cout << "Upside down transformation completed." << endl;
    return 0;
}

File: Python/2-Medium/00156._上下翻转二叉树/00156._上下翻转二叉树.py
from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def upsideDownBinaryTree(root: Optional[TreeNode]) -> Optional[TreeNode]:
    """
    上下翻转二叉树。

    Args:
    - root: Optional[TreeNode] - 二叉树的根节点。

    Returns:
    - Optional[TreeNode] - 翻转后的二叉树的根节点。

    示例:
    >>> 树 [1,2,3,4,5] 翻转后变为 [4,5,2,None,None,3,1]
    """
    # 递归基: 如果当前节点为空或者是叶子节点，直接返回当前节点
    if not root or (not root.left and not root.right):
        return root

    # 递归翻转左子树
    newRoot = upsideDownBinaryTree(root.left)

    # 当前节点的左子节点的右子节点设置为当前节点的右子节点
    root.left.right = root.right
    # 当前节点的左子节点的左子节点设置为当前节点
    root.left.left = root
    # 断开当前节点的左右子节点链接
    root.left = None
    root.right = None

    # 返回新的根节点
    return newRoot

# 辅助函数：构建二叉树
def buildTree(nodes, index=0):
    if index >= len(nodes) or nodes[index] is None:
        return None
    root = TreeNode(nodes[index])
    root.left = buildTree(nodes, 2 * index + 1)
    root.right = buildTree(nodes, 2 * index + 2)
    return root

# 辅助函数：层次遍历
def levelOrder(root):
    if not root:
        return []
    result, queue = [], [root]
    while queue:
        node = queue.pop(0)
        result.append(node.val)
        if node.left:
            queue.append(node.left)
        if node.right:
            queue.append(node.right)
    return result

# 构建并翻转二叉树
treeNodes = [1, 2, 3, 4, 5]
root = buildTree(treeNodes)
newRoot = upsideDownBinaryTree(root)

# 打印翻转后的二叉树层次遍历结果
print(levelOrder(newRoot))